### PR TITLE
Implement sparse random projections

### LINKED ISSUE #115: Implement random projections for covariance profiling
@mrohban I removed the random projections part in covariance profiling #114. 

I wondering whether we should have random projections as a separate feature transformation step? That's because if you have say, just 10 variables, it may be ok to have 45 covariance features. 

The disadvantage of that approach is that for larger feature sets, the dataframe can get too wide, take up memory, making it a pain to do random projections later. 

One mitigant for this is to do [sparse random projections](http://scikit-learn.org/stable/modules/random_projection.html#sparse-random-projection), which will be faster.

What do you think?

### PR DESCRIPTION
**Pull request by mcquin**
Resolves #115

### CODE REVIEW

**Comment by mcquin** [comment]
I'm mostly looking for feedback on implementation and style at this time. Also, is there anything I am missing (e.g., do I need to generate documentation)?

I'm concerned about the space complexity of `n_features * n_components`. It'd be great to do some bench marking before this gets merged.

File: R/sparse_random_projection.R [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#'    AreaShape_Length_DNA = c(2, 3, 1, 5),
+#'    Intensity_DNA = c(8, 20, 12, 32),
+#'    Texture_DNA = c(5, 2, 43, 13)
+#'  )
+#' variables <- c("AreaShape_Area_DNA", "AreaShape_Length_DNA", "Intensity_DNA", "Texture_DNA")
+#' sparse_random_projection(population, variables, 2)
+#'
+#' @importFrom magrittr %>%
+#' @importFrom magrittr %<>%
+#' @importFrom Matrix sparseMatrix
```
Comment (mcquin) [inline R/sparse_random_projection.R]: This probably belongs with `generate_component_matrix`.

File: R/sparse_random_projection.R [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+#' @importFrom stats rbinom
```
Comment (mcquin) [inline R/sparse_random_projection.R]: This probably belongs with `generate_component_matrix`.

File: R/sparse_random_projection.R [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#'
+#' @param n_features the dimensionality of the original space.
+#' @param n_components the dimensionality of the projected space.
+#' @param density the density of the sparse random matrix.
+#'
+#' @return A sparse random matrix of size \code{(n_features, n_components)}.
+#'
+#' @examples
+#' generate_component_matrix(500, 100, 0.3)
+generate_component_matrix <- function(n_features, n_components, density) {
```
Comment (mcquin) [inline R/sparse_random_projection.R]: `devtools::check()` yields the following error:

```
Error in generate_component_matrix(500, 100, 0.3) : 
  could not find function "generate_component_matrix"
```

**Review by shntnu — APPROVED** [review]
This was fun to review! Sorry it took so long!

**Pull request merged**